### PR TITLE
Prepare support for BlockBundle 4

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because BC.

## Changelog

```markdown
### Deprecated
- Passing a string as argument 1 when instantiating Sonata\DoctrineORMAdminBundle\Block\AuditBlockService 
```

### CODE REVIEW

File: src/Block/AuditBlockService.php [hunk 1]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
 
     /**
-     * @param string $name
+     * NEXT_MAJOR: Allow only Environment|EngineInterface for argument 1 and AuditReader for argument 2.
+     *
+     * @param Environment|EngineInterface|string $templatingOrDeprecatedName
+     * @param EngineInterface|AuditReader        $templatingOrAuditReader
      */
-    public function __construct($name, EngineInterface $templating, AuditReader $auditReader)
+    public function __construct($templatingOrDeprecatedName, $templatingOrAuditReader, ?AuditReader $auditReader = null)
```
Comment (greg0ire) [inline src/Block/AuditBlockService.php:42]:
```suggestion
    public function __construct($templatingOrDeprecatedName, object $templatingOrAuditReader, ?AuditReader $auditReader = null)
```

File: src/Block/AuditBlockService.php [hunk 2]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
+            parent::__construct($templatingOrDeprecatedName);
+
+            $this->auditReader = $templatingOrAuditReader;
+        } else {
+            throw new \TypeError(sprintf(
+                'Argument 2 passed to %s() must be either an instance of %s or preferably %s, %s given.',
+                __METHOD__,
+                EngineInterface::class,
+                AuditReader::class,
+                \is_object($templatingOrAuditReader) ? \get_class($templatingOrAuditReader) : \gettype($templatingOrAuditReader)
```
Comment (greg0ire) [inline src/Block/AuditBlockService.php:88]:
```suggestion
                \get_class($templatingOrAuditReader)
```

File: src/Block/AuditBlockService.php [hunk 3]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
+        } elseif ($templatingOrAuditReader instanceof AuditReader) {
+            if (!$templatingOrDeprecatedName instanceof Environment
+                && !$templatingOrDeprecatedName instanceof EngineInterface
+            ) {
+                throw new \TypeError(sprintf(
+                    'Argument 1 passed to %s() must be either an instance of %s or preferably %s, %s given.',
+                    __METHOD__,
+                    EngineInterface::class,
+                    Environment::class,
+                    \is_object($templatingOrAuditReader) ? \get_class($templatingOrAuditReader) : \gettype($templatingOrAuditReader)
```
Comment (greg0ire) [inline src/Block/AuditBlockService.php:75]:
```suggestion
                    \get_class($templatingOrAuditReader)
```

File: src/Block/AuditBlockService.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Block/AuditBlockService.php:88]:
```suggestion
                \is_object($templatingOrDeprecatedName) ? \get_class($templatingOrDeprecatedName) : \gettype($templatingOrDeprecatedName)
```

Was the right fix. Thanks for the catch.

File: src/Block/AuditBlockService.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Block/AuditBlockService.php:75]:
```suggestion
                \is_object($templatingOrDeprecatedName) ? \get_class($templatingOrDeprecatedName) : \gettype($templatingOrDeprecatedName)
```

Was the right fix. Thanks for the catch.

File: src/Block/AuditBlockService.php [hunk 4]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
+     * @param EngineInterface|AuditReader        $templatingOrAuditReader
      */
-    public function __construct($name, EngineInterface $templating, AuditReader $auditReader)
+    public function __construct($templatingOrDeprecatedName, object $templatingOrAuditReader, ?AuditReader $auditReader = null)
     {
-        parent::__construct($name, $templating);
+        if ($templatingOrAuditReader instanceof EngineInterface) {
+            @trigger_error(sprintf(
+                'Passing %s as argument 2 to %s() is deprecated since sonata-project/doctrine-orm-admin-bundle 3.x'
+                .' and will throw a \TypeError in version 4.0. You must pass an instance of %s instead.',
```
Comment (phansys) [inline src/Block/AuditBlockService.php:47]:
```suggestion
                .' and will throw a \TypeError exception in version 4.0. You must pass an instance of %s instead.',
```

File: src/Block/AuditBlockService.php [hunk 5]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
+        } elseif ($templatingOrAuditReader instanceof AuditReader) {
+            if (!$templatingOrDeprecatedName instanceof Environment
+                && !$templatingOrDeprecatedName instanceof EngineInterface
+            ) {
+                throw new \TypeError(sprintf(
+                    'Argument 1 passed to %s() must be either an instance of %s or preferably %s, %s given.',
+                    __METHOD__,
+                    EngineInterface::class,
+                    Environment::class,
+                    \is_object($templatingOrDeprecatedName) ? \get_class($templatingOrDeprecatedName) : \gettype($templatingOrDeprecatedName)
```
Comment (phansys) [inline src/Block/AuditBlockService.php:75]:
```suggestion
                    \is_object($templatingOrDeprecatedName) ? 'instance of '.\get_class($templatingOrDeprecatedName) : \gettype($templatingOrDeprecatedName)
```

File: src/Block/AuditBlockService.php [hunk 6]
```diff
@@ -33,13 +34,60 @@ class AuditBlockService extends AbstractBlockService
...
+                ));
+            }
+
+            parent::__construct($templatingOrDeprecatedName);
+
+            $this->auditReader = $templatingOrAuditReader;
+        } else {
+            throw new \TypeError(sprintf(
+                'Argument 2 passed to %s() must be either an instance of %s or preferably %s, %s given.',
```
Comment (phansys) [inline src/Block/AuditBlockService.php:84]:
```suggestion
                'Argument 2 passed to %s() must be either an instance of %s or preferably %s, instance of %s given.',
```

**Pull request merged**